### DISCUSSION TITLE
SpanProducer / SpanReader?

### DISCUSSION THREAD

**Discussion by hdost** [open-telemetry/opentelemetry-java#4029] [Q&A] [answered]
In the metrics side of the APIs there is a concept of a MetricProducer. However on the Tracing side no such thing exists. Is this something that would be likely implemented?

Example of the MetricReader https://github.com/open-telemetry/opentelemetry-java/blob/6ef3091cfb83946372da0d47d70acdf468c9e6c3/docs/sdk-configuration.md#configuring-the-sdk-in-a-framework

**Comment by jack-berg** [comment, marked as answer]
No, I don't think that's likely.

Metrics are aggregations of recorded measurements. The idea of a metric reader is that upon some signal, you read all the aggregated measurements in the form of metrics. If you have a pull based exporter, the signal is some consumer (such as a scraper) polling for metrics. If you have a push based exporter, the signal is a schedule which dictates periodic exports. 

Tracing doesn't involve aggregation. Also, there aren't any pull based mechanisms for exporting traces that I'm aware of. 

You could implement your own `SpanProcessor` that acted like a metric reader. It could accumulate spans in memory until some signal indicates that they should be exported in some fashion. Not sure why you'd want to do this though.

**Reply by hdost** [reply]
Yea the more I looked into it `SpanProcessor` seems the way.
